Seti 62  dynamic credentials

deliveroo-pauric opened the pull request:
Big update to how credentials are treated.
A "root" uuid will be stored as an environment variable. This key has the perms to add/delete/list client uuids, which are stored in redis.

Clients must use a uuid created by this user, at the /api_keys endpoint

This means restarting and hacky strings will no longer be required for authentication

deliveroo-pauric committed:
Test commit for debugging

tompave-roo committed:
Remove old file

tompave-roo committed:
rename the two auth middlewares, plus internal refactoring to fetch the key sources internally

tompave-roo committed:
update middleware references

deliveroo-pauric committed:
Update .env with new var, point to routemaster-client update in Gemfile

deliveroo-pauric committed:
Remove 3.1.0 routemaster-client gem

deliveroo-pauric committed:
Moved to seperate client/route auth middlewares, added controller for requesting new keys

deliveroo-pauric committed:
Tests

deliveroo-pauric committed:
Updated README

deliveroo-pauric committed:
Changed git:// to https:// to keep code climate happy

mezis commented on .env:
```diff
@@ -7,6 +7,9 @@ PUMA_THREADS=5
+ROUTEMASTER_ROOT_KEY=CHANGE-ME
+
 # credentials for allowed clients (comma-separated)
 ROUTEMASTER_CLIENTS=demo
```
I'm guessing this is no longer used?

deliveroo-pauric replied:
It wasn't used at all, I was exporting it locally. Thought it was a good way to point out that it needs to be set up for anyone doing a new install - the same way many projects do in OOTB configs

mezis replied:
I mean, the `ROUTEMASTER_CLIENTS` variables isn't used anymore, right?

deliveroo-pauric replied:
Oops, right you are.

mezis commented:
Looks good. One thing I _think_ this is missing is support for backwards compatibility / zero-downtime deploys. Maybe an option could be to prepopulate `Models::ClientToken` from `ROUTEMASTER_CLIENTS` at boot if the environment variable is present.

deliveroo-pauric commented:
Ok, could probably log a warning if it exists as well to make it clear it's deprecated. I'd have a slight concern that people would continue to use that method though, as it's still seen as working

gregbeech commented on Gemfile:
```diff
@@ -61,7 +61,7 @@ group :development do
 
   # testing against the client
-  gem 'routemaster-client'
+  gem 'routemaster-client', :git => 'https://github.com/deliveroo/routemaster-client', :branch => 'reset_client_connection'
```
`=>` arrows make Ruby 1.9+ sad 😢

gregbeech commented on spec/controllers/key_registration.rb:
```diff
@@ -0,0 +1,62 @@
+require 'spec_helper'
```
This file should probably be called `key_registration_spec.rb` to match the usual conventions.

gregbeech replied:
It _looks_ like it hasn't run as a result of the naming, as the code coverage indicates that the `KeyRegistration` class hasn't been hit.

mezis replied:
Yup. Also I don't think it'll run otherwise!

gregbeech commented on README.md:
```diff
@@ -280,8 +280,31 @@ All requests over non-SSL connections will be met with a 308 Permanent Redirect.
+#### Adding a client
+
+    >> POST /api_keys
+    >> { service_name: <string> }
```
Silly question maybe, but where does the key go? Is it passed in the payload, or returned in the response...?

gregbeech commented on README.md:
```diff
@@ -280,8 +280,31 @@ All requests over non-SSL connections will be met with a 308 Permanent Redirect.
+As `ROUTEMASTER_ROOT_KEY:<ignored>`
+Returns a 201 on success
+
+#### Deleteing a client
```
*Deleting

gregbeech commented on routemaster/controllers/key_registration.rb:
```diff
@@ -0,0 +1,35 @@
+      end
+
+      post '/api_keys', parse: :json do
+        if !data.has_key? "service_name"
```
`unless` might be cleaner than `if !`

gregbeech commented on routemaster/controllers/key_registration.rb:
```diff
@@ -0,0 +1,35 @@
+        end
+        new_key = Models::ClientToken.generate_api_key(data["service_name"])
+        status 201
+        {"new_key": new_key}.to_json
```
To be compliant with the [RFC 7231 § 6.3.2](https://tools.ietf.org/html/rfc7231#section-6.3.2) this should really include a `Location` header with the location of the new resource.

gregbeech commented on routemaster/models/client_token.rb:
```diff
@@ -0,0 +1,37 @@
+
+      def self.get_all
+        key_dump = {}
+        existing_keys = _redis.keys "#{PREFIX}*"
```
From https://redis.io/commands/keys:

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. If you're looking for a way to find keys in a subset of your keyspace, consider using SCAN or sets.

deliveroo-pauric replied:
I had came across that, but our use case here is fetching all keys at once, which we expect to be a relatively low number. I thought the point was moot, in that it's not quite the volume `select * from orders`. Additionally, it will very rarely need to be queried, so it essentially exists as a debugging interface

gregbeech replied:
The thing is that KEYS is proportional to the number of keys in Redis, not to the number of keys that match. So if a subscriber is backed up it could be a large number.

mezis replied:
In this case, I think you could use a Redis set as a data structure — then `SMEMBERS` here, `SADD` / `SREM` when adding/removing keys.

gregbeech commented on routemaster/models/client_token.rb:
```diff
@@ -0,0 +1,37 @@
+        while unique_key.nil? do
+          new_key = SecureRandom.uuid 
+          unique_key = new_key unless _redis.exists "#{PREFIX}#{new_key}"
+        end
```
This loops feels unnecessary to me--the change of actually duplicating a duplicate UUID is essentially zero, especially as we're only generating a small number. I'd change it to use `SecureRandom.hex(16)` though as UUIDs are guaranteed to be _unique_ but are not guaranteed to be _unguessable_ so shouldn't really be used as security tokens (the Ruby implementation is based on a secure generator so the implementation _happens_ to be safe here, but it's not guaranteed; you also only get about 122 bits of entropy rather than 128 as part of the layout is fixed to the v4 identifier and a few other bits).

gregbeech commented on routemaster/middleware/client_authentication.rb:
```diff
@@ -14,10 +15,8 @@ def call(env)
-          ENV.fetch('ROUTEMASTER_CLIENTS', '').split(','))
-        !! @_users.include?(username)
+      def _authenticate(uuid, _username)
+        !! Models::ClientToken.get_all.has_key?(uuid)
```
Materialising all keys and then checking for presence is going to be more expensive than checking for the presence of a single key on the server. I'd be inclined to make this a `exist?` method on `Models::ClientToken` so the implementation is hidden and then either cache the key set and query it locally, or just query a single key from Redis. This will be called a lot so we should make sure it's as performant as possible.

deliveroo-pauric committed:
Addressing the easy PR comments first

deliveroo-pauric committed:
Include Location header in 201 response

deliveroo-pauric committed:
Change the flow of redis storage

deliveroo-pauric commented:
After speaking with Tommaso, I used hashes instead of sets for the stoage. They're comparably cheap and mean we don't need to parsing by say tokenising strings to separate service name from uuid.

I had implemented a cache as well, but we talked about how it again convolutes the logic vs a cheap call to redis hget

gregbeech commented on routemaster/controllers/key_registration.rb:
```diff
@@ -0,0 +1,34 @@
+          halt 400, "expected 'service_name' key"
+        end
+        new_key = Models::ClientToken.generate_api_key(data["service_name"])
+        [201, {'Location': env["PATH_INFO"]}, {'new_key': new_key}.to_json]
```
Really the location should be the URL for `/api_keys/:key_name` assuming that's where the GET request would be (ideally it should also exist as an endpoint). Depends how fussed SETI is about this obeying correct HTTP semantics.

mezis replied:
SETI is very much fussed :)
If we're not exemplary on this it'll be hard to convince others! Thanks for singling this one out @gregbeech.

tompave-roo replied:
Not sure Routemaster should have a `GET /api_keys/:key_name` endpoint though. What use would it have?

What if instead of 201 here we return 200 or 204 and don't bother with the `Location` header?

gregbeech commented on routemaster/middleware/root_authentication.rb:
```diff
@@ -0,0 +1,32 @@
+        @protected_path === env["PATH_INFO"]
+      end
+
+      def _authenticate(uuid, _username)
```
Shouldn't `_username` really be `_password` here? I know it's not used so it doesn't matter, but if using a name rather than just `_` then it should probably be correct.

deliveroo-pauric replied:
It should be, but this is a long term bug that also exists in routemaster-client unfortunately

gregbeech commented on routemaster/models/client_token.rb:
```diff
@@ -0,0 +1,31 @@
+      end
+
+      def self.generate_api_key(service_name)
+        new_key = SecureRandom.uuid 
```
I think my comment may have got lost in a different change, but this should really be `.hex(16)` rather than `.uuid`. UUIDs are effectively guaranteed to be unique, but are not guaranteed to be unguessable. 2^128 bits of random data is both unique (with at least as strong a guarantee as a v4 UUID) and unguessable, so it a better choice for a security token or API key.

gregbeech commented on spec/controllers/key_registration_spec.rb:
```diff
@@ -0,0 +1,62 @@
+        {'CONTENT_TYPE' => 'application/json'}
+    end
+
+    let(:delete_key) do
```
Doesn't look like this is used anywhere?

deliveroo-pauric committed:
Further addressing PR comments

michael-groble commented on spec/middleware/client_authentication_spec.rb:
```diff
@@ -41,7 +42,7 @@ def call(env)
 
     it 'returns the username' do
```
`it 'returns the client uuid'` ?

deliveroo-pauric committed:
Correct wording of test description

deliveroo-pauric committed:
Changed new key response to 200, cleaned up some tests

deliveroo-pauric committed:
Updated readme, renamed two key reg vars

deliveroo-pauric commented:
Comments addressed (hopefully correctly :) )

mezis committed:
Merge branch 'master' into SETI-62__dynamic_credentials

mezis commented:
@deliveroo-pauric I took the liberty to merge master in.

mezis commented:
We're just missing an integration tests for the backwards-compatible loading from `ROUTEMASTER_CLIENTS`, I think.

deliveroo-pauric committed:
Added integration test for populating logins from env

deliveroo-pauric commented:
Test added

deliveroo-pauric committed:
Merge branch 'master' of github.com:deliveroo/routemaster into SETI-62__dynamic_credentials

deliveroo-pauric committed:
Merge branch 'master' into SETI-62__dynamic_credentials

deliveroo-pauric committed:
Merge branch 'SETI-62__dynamic_credentials' of github.com:deliveroo/routemaster into SETI-62__dynamic_credentials

mezis commented on spec/integration/client_spec.rb:
```diff
@@ -38,6 +36,10 @@
   let(:subscriber) { Routemaster::Models::Subscriber.find(uuid) }
   let(:topic) { Routemaster::Models::Topic.find('widgets') }
 
+  it 'populates logins from $ROUTEMASTER_CLIENTS at boot' do
```
That's not really an integration test, since you're checking implementation details (what is stored in Redis) instead of behaviour (can I use the API with that key).

Would you mind changing the test accordingly?

gregbeech commented on spec/integration/delivery_spec.rb:
```diff
@@ -94,14 +104,14 @@ def subscribe
     client.created('cats', 'https://example.com/cats/2')
     client.created('cats', 'https://example.com/cats/3')
-    processes.watch.wait_log %r(counter:events.added:2.*queue:demo)
+    processes.watch.wait_log %r(counter:events.added:2.*queue:#{uuid})
```
I'm not sure whether this is an issue or not so thought I'd at least ask. Is this `uuid` the security key for the client? If so we probably don't want to publish it in counters, but probably should use the service name.

deliveroo-pauric replied:
It is the security key, but it's not new in this PR, it's just that the key changed.

gregbeech reviewed (approved):
One question about the perf counters, otherwise LGTM.

mezis committed:
Merge branch 'master' into SETI-62__dynamic_credentials

mezis committed:
SETI-62: Clarification/cleanup

* Rename "API key" to "API token" throughout to reduce confusion
* Adds test coverage
* Updates README

munckymagik commented on README.md:
```diff
@@ -289,8 +289,41 @@ All requests over non-SSL connections will be met with a 308 Permanent Redirect.
+All allowed clients are stored in redis. These clients can be listed, added
+and deleted using the `ROUTEMASTER_ROOT_KEY` environment variable. All
+authentication is done with basic auth
+
```
Could we change this summary to something more similar to what has been written in the PR description?

E.g.

> All allowed clients are stored in Redis. A "root" uuid must be specified in the `ROUTEMASTER_ROOT_KEY` environment variable. This key has the perms to add/delete/list the client uuids. Clients must use a uuid created by this user, at the `/api_keys` endpoint.

... or something like that.

mezis replied:
Yup, that's much clearer. Thanks.

mezis replied:
I'll also update the "installing and configuring" section which still references the legacy`ROUTEMASTER_CLIENTS`.

mezis committed:
SETI-62: Converts auth middleware into Sinatra conditions

This enables better flexibility (multiple auth schemes per endpoint)
and makes controller addition order-independent.

mezis committed:
Merge branch 'master' into SETI-62__dynamic_credentials

mezis committed:
Testing leftover

mezis committed:
Comment clarification

mezis committed:
Makes Code Climate happier

mezis commented:
I've refactored this quite a bit (sorry) so that authentication is handled by [Sinatra conditions](https://stackoverflow.com/a/3560286/161487) instead of middleware.
The rationale is that we [need](https://github.com/deliveroo/routemaster-client/pull/21/files#diff-04c6e90faac2675aa89e2176d2eec7d8) some endpoints to be accessible by both authentication methods (root token or classic client token).

mezis committed:
Merge branch 'master' into SETI-62__dynamic_credentials

mezis committed:
Merge branch 'master' into SETI-62__dynamic_credentials

mezis committed:
Fixes Redis connection issues
